File: tests/reboot/documentation/task.ts
import { PartialMessage } from "@bufbuild/protobuf";
import {
  allow,
  Application,
  atMostOnce,
  Reboot,
  until,
  WorkflowContext,
  WriterContext,
} from "@reboot-dev/reboot";
import { OrderedMap } from "@reboot-dev/reboot-std/collections/ordered_map/v1";
import test from "node:test";
import { z } from "zod/v4";
import {
  CreateRequest,
  CreateResponse,
  CreateWorkflowRequest,
  CreateWorkflowResponse,
  OpenLaterRequest,
  OpenLaterResponse,
  OpenRequest,
  OpenResponse,
  PauseServiceRequest,
  PauseServiceResponse,
  ReadStateRequest,
  ReadStateResponse,
  StartPipelineRequest,
  StartPipelineResponse,
  TestTask,
  WelcomeEmailRequest,
  WelcomeEmailResponse,
} from "./task_rbt.js";

class TestTaskServicer extends TestTask.Servicer {
  authorizer() {
    return allow();
  }

  async create(
    context: WriterContext,
    request: CreateRequest
  ): Promise<PartialMessage<CreateResponse>> {
    return {};
  }

  static async startPipeline(
    context: WorkflowContext,
    request: StartPipelineRequest
  ): Promise<PartialMessage<StartPipelineResponse>> {
    const { taskId } = await TestTask.ref().loadS3Blob(context);
    return { taskId };
  }

  static async loadS3Blob(context: WorkflowContext) {
    return {};
  }

  static async readState(
    context: WorkflowContext,
    request: ReadStateRequest
  ): Promise<PartialMessage<ReadStateResponse>> {
    const snapshot = await TestTask.ref().read(context);
    return { snapshot };
  }

  static async pauseService(
    context: WorkflowContext,
    request: PauseServiceRequest
  ): Promise<PartialMessage<PauseServiceResponse>> {
    return {};
  }

  static async welcomeEmail(
    context: WorkflowContext,
    request: WelcomeEmailRequest
  ): Promise<PartialMessage<WelcomeEmailResponse>> {
    return {};
  }

  static async writeState(
    context: WorkflowContext,
    request: PauseServiceRequest
  ): Promise<PartialMessage<PauseServiceResponse>> {
    await TestTask.ref()
      .idempotently("Finally, increment the number of iterations")
      .write(context, async (state) => {
        state.iteration++;
      });

    // Won't execute again because same idempotency alias!
    await TestTask.ref()
      .idempotently("Finally, increment the number of iterations")
      .write(context, async (state) => {
        state.iteration++;
      });

    return {};
  }

  static async createWorkflow(
    context: WorkflowContext,
    request: CreateWorkflowRequest
  ): Promise<PartialMessage<CreateWorkflowResponse>> {
    await atMostOnce("something", context, async () => {
      return;
    });

    return {};
  }

  async open(
    context: WriterContext,
    request: OpenRequest
  ): Promise<PartialMessage<OpenResponse>> {
    this.state.customerName = request.customerName;

    const taskId = await this.ref().schedule().welcomeEmail(context);

    return { welcomeEmailTaskId: taskId };
  }

  async openLater(
    context: WriterContext,
    request: OpenLaterRequest
  ): Promise<PartialMessage<OpenLaterResponse>> {
    this.state.customerName = request.customerName;

    const now = new Date();
    const when = new Date(now.getTime() + 1000 * 60 * 60); // 1 hour from now

    const taskId = await this.ref().schedule({ when }).welcomeEmail(context);

    return { welcomeEmailTaskId: taskId };
  }

  static async twoEmails(
    context: WorkflowContext,
    request: OpenLaterRequest
  ): Promise<PartialMessage<OpenLaterResponse>> {
    const when = new Date();

    // task1 might complete after task2.
    const { task: task1 } = await TestTask.ref()
      .spawn({ when })
      .welcomeEmail(context);

    // task2 might complete before task1.
    const { task: task2 } = await TestTask.ref()
      .spawn({ when })
      .welcomeEmail(context);

    return { welcomeEmailTaskId: task1.taskId };
  }

  static async exampleOfUntil(context: WorkflowContext) {
    await until("Have messages", context, async () => {
      const { messages } = await TestTask.ref().read(context);
      return messages.length > 0;
    });

    const value = await until(
      "Value is stored",
      context,
      async () => {
        const index = OrderedMap.ref("someId");
        const { value } = await index.search(context, { key: "someKey" });
        return value.toJson();
      },
      { schema: z.json() }
    );
  }
}

// These tests are different from the other tests in this directory in that do not
// serve to test our system, instead they serve only to be examples in our
// documentation.

// Documentation examples notoriously go stale, as APIs can shift rapidly while
// docs are not pinned to individual versions. To prevent documentation drift and
// to make sure our users have a good experience, we strive to have every piece of
// code in our documentation be a tested, working piece of code that the user can
// `copy / paste` into their environment and it will just work.
test("task", async (t) => {
  let rbt: Reboot;
  t.before(async () => {
    rbt = new Reboot();
    await rbt.start();
    await rbt.up(new Application({ servicers: [TestTaskServicer] }));
  });
  t.after(async () => {
    await rbt.stop();
  });
  await t.test("create", async (t) => {
    const context = rbt.createExternalContext("test");

    const [task] = await TestTask.idempotently().create(context, "task");

    const {
      task: { taskId },
    } = await task.spawn().createWorkflow(context, {});

    await TestTask.CreateWorkflowTask.retrieve(context, { taskId });
  });
});

File: tests/reboot/documentation/task.py
import asyncio
from datetime import datetime, timedelta
from reboot.aio.applications import Application
from reboot.aio.auth.authorizers import allow
from reboot.aio.contexts import WorkflowContext, WriterContext
from reboot.aio.external import InitializeContext
from reboot.aio.workflows import until
from reboot.std.collections.v1.sorted_map import SortedMap
from tests.reboot.documentation.task_rbt import (
    OpenLaterRequest,
    PauseServiceRequest,
    PauseServiceResponse,
    ReadStateRequest,
    ReadStateResponse,
    TestTask,
    TwoEmailsRequest,
)


class TestTaskServicer(TestTask.Servicer):

    def authorizer(self):
        return allow()

    async def open_later(
        self,
        context: WriterContext,
        request: OpenLaterRequest,
    ):
        self.state.customerName = request.customerName

        now = datetime.now()
        when = now + timedelta(hours=1)

        await self.ref().schedule(
            when=when,
        ).welcome_email(context)

    @classmethod
    async def two_emails(
        cls,
        context: WorkflowContext,
        request: TwoEmailsRequest,
    ):
        now = datetime.now()

        # This task might complete before or after next.
        _ = await TestTask.ref().spawn(
            when=now + timedelta(hours=1),
        ).welcome_email(context)

        # This task might complete before or after previous.
        _ = await TestTask.ref().spawn(
            when=now + timedelta(hours=1),
        ).welcome_email(context)

    @classmethod
    async def read_state(
        cls,
        context: WorkflowContext,
        request: ReadStateRequest,
    ) -> ReadStateResponse:
        snapshot = await TestTask.ref().read(context)
        return ReadStateResponse(
            snapshot=snapshot,
        )

    @classmethod
    async def pause_service(
        cls,
        context: WorkflowContext,
        request: PauseServiceRequest,
    ) -> PauseServiceResponse:

        async def increment(state):
            state.iteration += 1

        await TestTask.ref().idempotently(
            "Finally, increment the number of iterations",
        ).write(
            context,
            increment,
        )

        # Won't execute again because same idempotency alias!
        await TestTask.ref().idempotently(
            "Finally, increment the number of iterations",
        ).write(
            context,
            increment,
        )

        async def have_messages():
            state = await TestTask.ref().read(context)
            return len(state.messages) > 0

        await until("Have messages", context, have_messages)

        async def value_is_stored():
            map = SortedMap.ref("someId")
            response = await map.get(context, key="someKey")
            return response.value if response.HasField("value") else False

        value = await until(
            "Value is stored",
            context,
            value_is_stored,
            type=bytes,
        )

        del value

        return PauseServiceResponse()


async def initialize(context: InitializeContext):
    await TestTask.create(context, "task")


# These tests are different from the other tests in this directory in that do not
# serve to test our system, instead they serve only to be examples in our
# documentation.


# Documentation examples notoriously go stale, as APIs can shift rapidly while
# docs are not pinned to individual versions. To prevent documentation drift and
# to make sure our users have a good experience, we strive to have every piece of
# code in our documentation be a tested, working piece of code that the user can
# `copy / paste` into their environment and it will just work.
async def main():
    application = Application(
        servicers=[TestTaskServicer],
    )
    await application.run()


if __name__ == '__main__':
    asyncio.run(main())
